extensions/wikia/AdEngine/ResourceLoaders/ResourceLoaderAdEngineBase.php

```diff
@@ -0,0 +1,180 @@
+	 * @param $path
+	 * @return string
+	 */
+	protected function fetchLocalScript( $path ) {
+		return file_get_contents( $path );
+	}
+
+	/**
+	 * @param $scriptValue
+	 * @return mixed
+	 */
+	protected function getInlineScript( $scriptValue ) {
+		return $scriptValue;
+	}
+
+	/**
+	 * Merges sripts into one
+	 * @param array $scriptItems
+	 * @return bool|string
+	 */
+	protected function generateData( $scriptItems ) {
+
+		$scriptsToMerge = [];
+		foreach ( $scriptItems as $scriptItem ) { /* @var ResourceLoaderScript $scriptItem */
+
+			$script = false;
+
+			if ( $scriptItem->getType() == ResourceLoaderScript::TYPE_REMOTE ) {
+				$script = $this->fetchRemoteScript( $scriptItem->getValue() );
+			}
+			elseif ( $scriptItem->getType() == ResourceLoaderScript::TYPE_LOCAL ) {
+				$script = $this->fetchLocalScript( $scriptItem->getValue() );
+			}
+			elseif ( $scriptItem->getType() == ResourceLoaderScript::TYPE_INLINE ) {
+				$script = $this->getInlineScript( $scriptItem->getValue() );
+			}
+
+			if ( $script === false ) {
+				return false;
+			}
+
+			$scriptsToMerge[] = $script;
+		}
+
+		return join(PHP_EOL, $scriptsToMerge);
+	}
+
+	/**
+	 * Current timestamp generated only once per script execution
+	 * @return int
+	 */
+	protected function getCurrentTimestamp() {
+		static $now;
+		if (!$now) {
+			$now = time();
+		}
+		return $now;
+	}
+
+	/**
+	 * Get data from cache / generate new data if TTL is less than time()
+	 * @return array|bool|Object
+	 */
+	protected function getData() {
+		global $wgMemc;
+
+		if (static::$localCache[get_class($this)]) {
+			return static::$localCache[get_class($this)];
+		}
+
+		$now = $this->getCurrentTimestamp();
+
+		$memKey = wfSharedMemcKey('adengine', get_class($this) . __FUNCTION__, static::CACHE_BUSTER);
+
+		$cached = $wgMemc->get($memKey);
+		if (is_array($cached) && $cached['ttl'] > $now) {
+			// Cache hit!
+			static::$localCache[get_class($this)] = $cached;
+			return $cached;
+		}
+
+		// Cache miss, need to re-download the scripts
+		$generated = $this->generateData( $this->getScripts() );
+
+		if ($generated === false) {
+			// HTTP request didn't work
+
+			if (is_array($cached)) {
+				// Oh, we still have the thing cached
+				// Let's use the script for the next a few minutes
+
+				$cached['ttl'] = $now + static::TTL_GRACE;
+				$wgMemc->set($memKey, $cached);
+				static::$localCache[get_class( $this )] = $cached;
+				return $cached;
+			}
+			$data = $this->getFallbackDataWhenRequestFails();
+			$wgMemc->set($memKey, $data);
+
+			\Wikia\Logger\WikiaLogger::instance()
+				->warning( 'ResourceLoaderAdEngine - failed to fetch new data',
+					['class' => get_class( $this )] );
+
+			static::$localCache[get_class( $this )] = $data;
+			return $data;
+		}
+
+		$data = [
+			'script' => $generated,
+			'modTime' => $now,
+			'ttl' => $now + static::TTL_SCRIPTS,
+		];
+
+		if ( md5($data['script']) != md5($cached['script']) ) {
```

Comment:
It's a good practice in PHP to use `===` and `!==` operators but those methods generates always a string, so we're fine here IMHO.